### ISSUE TITLE
Notifications- Remaining activities

### ISSUE THREAD

**Issue by p-sad** [Joystream/pioneer#933]
||Working group|Worker|Member|Proposal|Forum
| - | - | - | - | - | -
|ApplicationWithdrawnEvent|:heavy_check_mark:||:heavy_check_mark:!||
|AppliedOnOpeningEvent|:heavy_check_mark:|:heavy_check_mark:!|:heavy_check_mark:!||
|BudgetSetEvent|:heavy_check_mark:||||
|BudgetSpendingEvent|:heavy_check_mark:||||
|InitialInvitationBalanceUpdatedEvent|||||
|InitialInvitationCountUpdatedEvent|||||
|InvitesTransferredEvent|||||
|LeaderInvitationQuotaUpdatedEvent|||||
|LeaderSetEvent|||||
|LeaderUnsetEvent|||||
|MemberAccountsUpdatedEvent|||||
|MemberInvitedEvent|||||
|MemberProfileUpdatedEvent|||||
|MemberVerificationStatusUpdatedEvent|||||
|MembershipBoughtEvent|||||
|MembershipPriceUpdatedEvent|||||
|NewMissedRewardLevelReachedEvent|||||
|OpeningAddedEvent|:heavy_check_mark:||||
|OpeningCanceledEvent|:heavy_check_mark:||||
|OpeningFilledEvent|:heavy_check_mark:|x!|x!||
|ProposalCancelledEvent||||:heavy_check_mark:|
|ProposalCreatedEvent||||:heavy_check_mark:|
|ProposalDecisionMadeEvent||||:heavy_check_mark:|
|ProposalDiscussionPostCreatedEvent||||:heavy_check_mark:|
|ProposalDiscussionPostDeletedEvent||||:heavy_check_mark:|
|ProposalDiscussionPostUpdatedEvent||||:heavy_check_mark:|
|ProposalDiscussionThreadModeChangedEvent||||:heavy_check_mark:|
|ProposalExecutedEvent||||:heavy_check_mark:|
|ProposalStatusUpdatedEvent||||:heavy_check_mark:|
|ProposalVotedEvent|||as voter!|:heavy_check_mark:|
|ReferralCutUpdatedEvent|||||
|RewardPaidEvent|||||
|StakeDecreasedEvent|:heavy_check_mark:|:heavy_check_mark:|w:heavy_check_mark:||
|StakeIncreasedEvent|:heavy_check_mark:|:heavy_check_mark:|w:heavy_check_mark:||
|StakeSlashedEvent|:heavy_check_mark:|:heavy_check_mark:|w:heavy_check_mark:||
|StakingAccountAddedEvent|||||
|StakingAccountConfirmedEvent|||||
|StakingAccountRemovedEvent|||||
|StatusTextChangedEvent|:heavy_check_mark:||||
|TerminatedLeaderEvent|:heavy_check_mark:|:heavy_check_mark:|w:heavy_check_mark:||
|TerminatedWorkerEvent|:heavy_check_mark:|:heavy_check_mark:|w:heavy_check_mark:||
|WorkerExitedEvent|:heavy_check_mark:|:heavy_check_mark:|w:heavy_check_mark:||
|WorkerRewardAccountUpdatedEvent||:heavy_check_mark:|w:heavy_check_mark:||
|WorkerRewardAmountUpdatedEvent|??|:heavy_check_mark:|w:heavy_check_mark:||
|WorkerRoleAccountUpdatedEvent|||||
|WorkerStartedLeavingEvent||:heavy_check_mark:|w:heavy_check_mark:||
|_PostDeletedEvent_|||||:heavy_check_mark:
|_PostModeratedEvent_|||||
|_PostAddedEvent_|||x!||:heavy_check_mark:
|_PostTextUpdatedEvent_|||||:heavy_check_mark:
|_PostReactedEvent_|||x!?, see below||x
|_ThreadMovedEvent_|||||x
|_ThreadCreatedEvent_|||||:heavy_check_mark:
|_ThreadDeletedEvent_|||||x
|_ThreadModeratedEvent_|||||
|_ThreadMetadataUpdatedEvent_|||||
|_CategoryCreatedEvent_|||||:heavy_check_mark:
|_CategoryArchivalStatusUpdatedEvent_|||||x
|_CategoryDeletedEvent_|||||x
|_CategoryMembershipOfModeratorUpdatedEvent_|||||x
|_CategoryStickyThreadUpdateEvent_|||||x

I'm pretty sure the table now contains every event that currently exists in the schema. Legend:

* x - Needed in this notifications group
* :heavy_check_mark: - needed and has a component already
* ! - means that a notification needs to appear in a group, but the event has no straightforward way to reference by entity id. e.g.: `ApplicationWithdrawnEvent` can only be referenced by an application ID, not by member ID directly.
* ? - more complex explanation, outlined below.
* w - in member notifications, means that the query `where` uses a `worker { id_in }` instead of member id to find them

Proposal activities work extremely well when querying for the `Event` interface; it works this well thanks in part to that query not using any other parameters (unlike e.g. WG activities filtering by `group`). Forum activities would greatly benefit from the same treatment because they're also not tied to any parameter, but for some reason are missing from the `Event` interface. Would work great if added.

Because Worker/Working Group activities always use the same parameter in queries (`worker` and `group` respectively), there is a case to be made for creating two additional interfaces - besides `Event`

* that would each implement a worker or a group parameter that could be used in a single `where` clause. This would be useful for fetching the actual _n_ most recent events for a given group/worker, instead of an approximation.

Member notifications are probably the trickiest to add. Most of them are just worker activities, and can probably be queried using a member's workers list, but the rest don't look like they were designed to be implemented. Understandably, most events reference a member by their role in the event (e.g. as `voter` in `ProposalVotedEvent`) so a common interface cannot be used. In others the member can't be referenced directly (like post added and opening application events). It seems like out of proposal events, only "proposal voted" is considered as a notification, though that might just be referencing a forum poll instead. It seems like they could use a redesign/overhaul.

Activities with missing events & question marks:

* Events for updating forum category title and description don't seem to exist in the schema
* Event for updating forum thread title doesn't seem to exist in the schema
* User being mentioned in a post can probably only be found through `PostAddedEvent.post.repliesTo` which is unfeasible to implement. There's no event that just says one's post was replied to.
* User being mentioned in a post can only be found via matching post text, which is unfeasible to do
* Workers hired list in `OpeningFilledEvent` does not conform with the way all other worker activities are queried (that's for a "you have been hired" activity)
* Worker's "applied on opening" activity seems incredibly contrived to query, seeing that you'd need to actually query `Worker.application.createdInEvent` to find the event itself
* ?? - there is an activity saying that group reward was increased, but AFAIK worker rewards are adjusted per worker
* ? - `PostReactedEvent` only references the reacting member directly, the member reacted to can only be found in post fields (and there are activities for both reacting and being reacted to)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201703952879841/1201708908246026) by [Unito](https://www.unito.io)

**Comment by p-sad** [comment]
Previous version of the description:

| Event | Component | TS type | Query+cast | Mocks |
| - | - | - | - | -|
|OpeningFilledEvent|~|x|~|x
| BudgetSetEvent |~|x|x|x
|WorkerRewardAccountUpdatedEvent
|WorkerRewardAmountUpdatedEvent
|ProposalCancelledEvent
|ProposalCreatedEvent
|ProposalDecisionMadeEvent
|ProposalDiscussionPostCreatedEvent
|ProposalDiscussionPostDeletedEvent
|ProposalDiscussionPostUpdatedEvent
|ProposalDiscussionThreadCreatedEvent
|ProposalDiscussionThreadModeChangedEvent
|ProposalExecutedEvent
|ProposalStatusUpdatedEvent
|ProposalVotedEvent
|PostAddedEvent||||x
|PostTextUpdatedEvent||||x
|PostDeletedEvent
|PostReactedEvent
|~LeaderSetEvent~
|~LeaderUnsetEvent~

x means done, ~ means partly done or challenge encountered
~Strikethrough~ means activity is no longer in the design but might be good to still keep it in mind

**Comment by bedeho** [comment]
Is this part of Olympia?

**Comment by thesan** [comment]
Yes I think we should assess which notifications are missing/not fully implemented for Olympia
